homeassistant/helpers/entity.py

```diff
@@ -775,12 +781,22 @@ def _friendly_name_internal(self) -> str | None:
         return f"{device_name} {name}" if device_name else name
 
     @callback
-    def _async_generate_attributes(self) -> tuple[str, dict[str, Any]]:
-        """Calculate state string and attribute mapping."""
+    def _async_calculate_state(
+        self,
+    ) -> tuple[str, dict[str, Any], Mapping[str, Any] | None, Mapping[str, Any]]:
```

Comment:
What about using a named tuple for return value? How performant is that?

Reply:
I extended @bdraco's benchmark:

```py
from collections.abc import Mapping
from dataclasses import dataclass
import timeit
from typing import Any,NamedTuple

empty_dict = {}


@dataclass(frozen=True, slots=True)
class CalculatedState:
    """Container with state and attributes.
    Returned by Entity._async_calculate_state.
    """

    state: str
    # The union of all attributes, after overriding with entity registry settings
    attributes: dict[str, Any]
    # Capability attributes returned by the capability_attributes property
    capability_attributes: Mapping[str, Any] | None
    # Attributes which may be overridden by the entity registry
    shadowed_attributes: Mapping[str, Any]


class CalculatedStateTuple(NamedTuple):
    state: str
    # The union of all attributes, after overriding with entity registry settings
    attributes: dict[str, Any]
    # Capability attributes returned by the capability_attributes property
    capability_attributes: Mapping[str, Any] | None
    # Attributes which may be overridden by the entity registry
    shadowed_attributes: Mapping[str, Any]


def test_dataclass():
    # mypy fails to narrow the type in the above if statement above
    return CalculatedState("state", empty_dict, None, empty_dict)


def test_named_tuple():
    return CalculatedStateTuple("state", empty_dict, None, empty_dict)


def test_with_tuple():
    return ("state", empty_dict, None, empty_dict)


dataclass_time = timeit.timeit(test_dataclass)
named_tuple_time = timeit.timeit(test_named_tuple)
tuple_time = timeit.timeit(test_with_tuple)


print(f"Time for tuple: {tuple_time}")
print(f"Time for named tuple: {named_tuple_time}")
print(f"Time for dataclass: {dataclass_time}")
```

The conclusion is that creating a named tuple instance is almost as slow as creating a frozen dataclass instance:
```
$ python3 benchmark_containers.py
Time for tuple: 0.05015225103124976
Time for named tuple: 0.2931557139963843
Time for dataclass: 0.6470301699591801
```

Reply:
As discussed on Discord, I also tried regular and type dicts, which both need a very similar amount of time to the named tuple:
```
python3 benchmark_containers.py
Time for dataclass: 0.6543728460092098
Time dict: 0.34074735996546224
Time for named tuple: 0.2823609299957752
Time for tuple: 0.0498239989974536
Time for typed dict: 0.28705969301518053
```